mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -289,17 +301,28 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
-
-    val sentences: RDD[Array[Int]] = words.mapPartitions { iter =>
+    // each partition is a collection of sentences, will be translated into arrays of Index integer
+    val sentences: RDD[Array[Int]] = dataset.mapPartitions { sentenceIter =>
       new Iterator[Array[Int]] {
-        def hasNext: Boolean = iter.hasNext
+        var wordIter: Iterator[String] = null
+
+        def hasNext: Boolean = sentenceIter.hasNext || (wordIter != null && wordIter.hasNext)
 
         def next(): Array[Int] = {
           val sentence = ArrayBuilder.make[Int]
           var sentenceLength = 0
-          while (iter.hasNext && sentenceLength < MAX_SENTENCE_LENGTH) {
-            val word = bcVocabHash.value.get(iter.next())
-            word match {
+          // do translation of each word into its index in the vocabulary,
```

Comment:
Good point; I'm not sure if it was for efficiency or simply to copy the C++ implementation. Hm, I don't suppose we have benchmarks here. I'm OK with keeping the longer more efficient implementation. but now that input sentence boundaries matter, this can still be simplified. This part can still be used:

```
// Each input sentence will produce 1 or more Array[Int], so flatMapPartitions
dataset.flatMapPartitions { sentenceIter => 
  // Each sentence will map to 1 or more Array[Int], so map
  sentenceIter.map { sentence =>
    ...
  }
}
```

That is you can still `flatMapPartitions` over `Iterator`s for each sentence to join them without trying to manage both the `Iterator` over sentences and words.